ada/language/ast.py

```diff
@@ -2804,7 +2804,22 @@ def real_type(typ=T.BaseTypeDecl.entity):
         # Compute the canonical types and discard the metadata fields for a
         # more robust comparison.
         canon_prim_type = Var(prim_type._.canonical_type._.without_md)
-        canon_typ = Var(typ.then(lambda t: t.canonical_type.without_md))
+
+        canon_typ = Var(Cond(
+            typ.is_null,
+            No(BaseTypeDecl.entity),
+
+            # Don't call canonical_type on classwide types, as it returns a
+            # non-classwide type for them, which could make the comparison
+            # below incorrectly succeed.
+            # TODO: Fixing canonical_type directly would probably be better,
```

Comment:
As discussed live: I'd like to have an idea of how hard it would be to do the real fix, rather than a workaround here

Reply:
Actual fix done!